### PR TITLE
Add support for defining simple custom utilities in CSS via `@utility`

### PR DESCRIPTION
**Pull request by thecrypticace**
This PR allows you to add custom utilities to your project via the new `@utility` rule.

For example, given the following:
```css
@utility text-trim {
  text-box-trim: both;
  text-box-edge: cap alphabetic;
}
```

A new `text-trim` utility is available and can be used directly or with variants:
```html
<div class="text-trim">...</div>
<div class="hover:text-trim">...</div>
<div class="lg:dark:text-trim">...</div>
```

If a utility is defined more than one time the latest definition will be used:
```css
@utility text-trim {
  text-box-trim: both;
  text-box-edge: cap alphabetic;
}

@utility text-trim {
  text-box-trim: both;
  text-box-edge: cap ideographic;
}
```

Then using `text-trim` will produce the following CSS:
```css
.text-trim {
  text-box-trim: both;
  text-box-edge: cap ideographic;
}
```

You may also override specific existing utilities with this — for example to add a `text-rendering` property to the `text-sm` utility:
```css
@utility text-sm {
  font-size: var(--font-size-sm, 0.875rem);
  line-height: var(--font-size-sm--line-height, 1.25rem);
  text-rendering: optimizeLegibility;
}
```

Though it's preferred, should you not need to add properties, to override the theme instead.

Lastly, utilities with special characters do not need to be escaped like you would for a class name in a selector:
```css
@utility push-1/2 {
  right: 50%;
}
```

We do however explicitly error on certain patterns that we want to reserve for future use, for example `push-*` and `push-[15px]`.

### CODE REVIEW

File: packages/tailwindcss/src/index.ts [hunk 1]
```diff
@@ -65,6 +67,31 @@ export function compile(
       return
     }
 
+    // Collect custom `@utility` at-rules
+    if (node.selector.startsWith('@utility ')) {
+      let name = node.selector.slice(9).trim()
+
+      if (!/^[a-z][a-zA-Z0-9/%._-]*$/.test(name)) {
+        throw new Error(
+          `The utility [${name}] has an invalid name. Only use alphanumeric characters.`,
```
Comment (adamwathan) [inline packages/tailwindcss/src/index.ts:78]: > `@utility ${name}` defines an invalid utility name. Utilities should be alphanumeric and start with a lowercase letter.

Comment (adamwathan) [inline packages/tailwindcss/src/index.ts:78]:
```suggestion
          `\`@utility ${name}\` defines an invalid utility name. Utilities should be alphanumeric and start with a lowercase letter.`,
```

File: packages/tailwindcss/src/index.ts [hunk 2]
```diff
@@ -65,6 +69,31 @@ export function compile(
...
+      let name = node.selector.slice(9).trim()
+
+      if (!IS_VALID_UTILITY_NAME.test(name)) {
+        throw new Error(
+          `The utility [${name}] has an invalid name. Only use alphanumeric characters.`,
+        )
+      }
+
+      if (node.nodes.length === 0) {
+        throw new Error(`The utility [${name}] needs at least one property.`)
```
Comment (adamwathan) [inline packages/tailwindcss/src/index.ts:83]:
```suggestion
        throw new Error(`\`@utility ${name}\` is empty. Utilities should include at least one property.`)
```

File: packages/tailwindcss/src/utilities.ts [hunk 3]
```diff
@@ -73,35 +69,12 @@ export class Utilities {
...
   }
 
-  keys() {
-    return this.utilities.keys()
-  }
-
-  entries() {
-    return this.utilities.entries()
+  keys(kind: 'static' | 'functional') {
+    return Array.from(this.utilities.keys()).filter((key) => this.has(key, kind))
```
Comment (philipp-spiess) [inline packages/tailwindcss/src/utilities.ts:73]: I’m not sure how often this is called (if at all?) but we could probably avoid an array allocation here if we want

File: packages/tailwindcss/src/index.ts [hunk 4]
```diff
@@ -65,6 +69,33 @@ export function compile(
...
+          `\`@utility ${name}\` is empty. Utilities should include at least one property.`,
+        )
+      }
+
+      customUtilities.push((designSystem) => {
+        designSystem.utilities.static(name, (candidate) => {
+          if (candidate.negative) return
+          return structuredClone(node.nodes)
+        })
+      })
```
[the remark this answers is not in the document]
Comment (thecrypticace) [reply packages/tailwindcss/src/index.ts:93]: we explicitly support this — see the tests.

File: packages/tailwindcss/src/utilities.ts (hunk 3 above)
Comment (thecrypticace) [reply packages/tailwindcss/src/utilities.ts:73]: yeah we can optimize this — was going to do that in a followup PR.  This is only used in intellisense so it matters less.

File: packages/tailwindcss/src/utilities.test.ts [hunk 5]
```diff
@@ -14946,3 +14947,180 @@ test('@container', () => {
     ]),
   ).toEqual('')
 })
+
+describe('custom utilities', () => {
```
[the remark this answers is not in the document]
Comment (thecrypticace) [reply packages/tailwindcss/src/utilities.test.ts:14951]: done

File: packages/tailwindcss/src/compile.ts [hunk 6]
```diff
@@ -121,10 +121,20 @@ export function compileAstNodes(rawCandidate: string, designSystem: DesignSystem
     // operator because if the `candidate.root` didn't exist, `parseCandidate`
     // would have returned `null` and we would have returned early resulting
     // in not hitting this code path.
```
[the remark this answers is not in the document]
Comment (thecrypticace) [reply packages/tailwindcss/src/compile.ts]: Done

File: packages/tailwindcss/src/utilities.ts [hunk 7]
```diff
@@ -29,38 +28,35 @@ type SuggestionDefinition =
...
-
-  has(name: string) {
-    return this.utilities.has(name)
+    this.arbitraryFn = compileFn
   }
 
-  get(name: string | symbol) {
-    return this.utilities.get(name)
+  has(name: string, kind: 'static' | 'functional') {
+    return this.utilities.has(name) && this.utilities.get(name).some((fn) => fn.kind === kind)
```
Comment (RobinMalfait) [inline packages/tailwindcss/src/utilities.ts:55]: Q: Does it make sense to store the utilities on a per-kind basis? This way we can do the lookup once and don't need to do the `.some()` part ever 🤔

File: packages/tailwindcss/src/utilities.ts (hunk 3 above)
Comment (thecrypticace) [reply packages/tailwindcss/src/utilities.ts]: Updated

File: packages/tailwindcss/src/utilities.ts (hunk 7 above)
Comment (thecrypticace) [reply packages/tailwindcss/src/utilities.ts:55]: We did that first but:
- It looked awful
- You lose potential ordering between multiple definitions of utilities

**Review by philipp-spiess — APPROVED** [review]

**Review by adamwathan — APPROVED** [review]

**Pull request merged**